src/xbacklight.cc

```diff
@@ -0,0 +1,443 @@
+				      1, (unsigned char *)&value);
+}
+
+int
+_xbacklight (int argc, char **argv)
+{
+    char    *dpy_name = NULL;
+    op_t    op = Get;
+    int	    value = 0;
+    int	    i;
+    int	    total_time = 200;	/* ms */
+    int	    steps = 20;
+
+    xcb_connection_t *conn;
+    xcb_generic_error_t *error;
+
+    xcb_randr_query_version_cookie_t ver_cookie;
+    xcb_randr_query_version_reply_t *ver_reply;
+
+    xcb_intern_atom_cookie_t backlight_cookie[2];
+    xcb_intern_atom_reply_t *backlight_reply;
+
+    xcb_screen_iterator_t iter;
+
+    program_name = argv[0];
+
+    for (i = 1; i < argc; i++)
+    {
+	if (!strcmp (argv[i], "-display") || !strcmp ("-d", argv[i]))
+	{
+	    if (++i >= argc) missing_arg (argv[i-1]);
+	    dpy_name = argv[i];
+	    continue;
+	}
+	if (!strcmp (argv[i], "-set") || !strcmp (argv[i], "="))
+	{
+	    if (++i >= argc) missing_arg (argv[i-1]);
+	    op = Set;
+	    value = atoi (argv[i]);
+	    continue;
+	}
+	if (argv[i][0] == '=' && isdigit (argv[i][1]))
+	{
+	    op = Set;
+	    value = atoi (argv[i] + 1);
+	    continue;
+	}
+	if (!strcmp (argv[i], "-inc") || !strcmp (argv[i], "+"))
+	{
+	    if (++i >= argc) missing_arg (argv[i-1]);
+	    op = Inc;
+	    value = atoi (argv[i]);
+	    continue;
+	}
+	if (argv[i][0] == '+' && isdigit (argv[i][1]))
+	{
+	    op = Inc;
+	    value = atoi (argv[i] + 1);
+	    continue;
+	}
+	if (!strcmp (argv[i], "-dec") || !strcmp (argv[i], "-"))
+	{
+	    if (++i >= argc) missing_arg (argv[i-1]);
+	    op = Dec;
+	    value = atoi (argv[i]);
+	    continue;
+	}
+	if (argv[i][0] == '-' && isdigit (argv[i][1]))
+	{
+	    op = Dec;
+	    value = atoi (argv[i] + 1);
+	    continue;
+	}
+	if (!strcmp (argv[i], "-get") || !strcmp (argv[i], "-g"))
+	{
+	    op = Get;
+	    continue;
+	}
+	if (!strcmp (argv[i], "-time"))
+	{
+	    if (++i >= argc) missing_arg (argv[i-1]);
+	    total_time = atoi (argv[i]);
+	    continue;
+	}
+	if (!strcmp (argv[i], "-steps"))
+	{
+	    if (++i >= argc) missing_arg (argv[i-1]);
+	    steps = atoi (argv[i]);
+	    continue;
+	}
+	if (!strcmp (argv[i], "-help") || !strcmp (argv[i], "-?"))
+	{
+	    usage (0);
+	}
+	if (!strcmp (argv[i], "-version"))
+	{
+	    puts ("0.1.0");
+	    exit (0);
+	}
+	fprintf(stderr, "%s: unrecognized argument '%s'\n",
+		program_name, argv[i]);
+	usage (1);
+    }
+    conn = xcb_connect (dpy_name, NULL);
+    ver_cookie = xcb_randr_query_version (conn, 1, 2);
+    ver_reply = xcb_randr_query_version_reply (conn, ver_cookie, &error);
+    if (error != NULL || ver_reply == NULL) {
+	int ec = error ? error->error_code : -1;
+	fprintf (stderr, "RANDR Query Version returned error %d\n", ec);
+	exit (1);
+    }
+    if (ver_reply->major_version != 1 ||
+	ver_reply->minor_version < 2) {
+	fprintf (stderr, "RandR version %d.%d too old\n",
+		 ver_reply->major_version, ver_reply->minor_version);
+	exit (1);
+    }
+    free (ver_reply);
+
+    backlight_cookie[0] = xcb_intern_atom (conn, 1, strlen("Backlight"), "Backlight");
```

Comment:
Factor this function into two - the first part parses arguments, the second part drives the xcb API. Call the 2nd part directly from your node driver rather than synthesizing a command line string.